Readme.md
# GPGit

![gpgit.png](img/gpgit.png)

# Introduction
As we all know, today more than ever before, it is crucial to be able to trust our computing environments. One of the main difficulties that package maintainers of GNU/Linux distributions face, is the difficulty to verify the authenticity and the integrity of the source code. With GPG signatures it is possible for packagers to verify source code releases quickly and easily.

#### Overview of the required tasks:
* Create and/or use a **[4096-bit RSA/Ed25519 ECC keypair][1]** for the file signing
* Use a **[strong, unique, secret passphrase][2]** for the key
* Upload the public key to a **[key server][3]** and **[publish the full fingerprint][4]**
* **[Sign][5]** every new Git **[commit][6]** and **[tag][7]**
* Create **[signed][8], [compressed release archives][9]**
* Upload a **[strong message digest][10]** of the archive
* Configure **[HTTPS][11]** for your download server

<a href="https://www.buymeacoffee.com/nicohood" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png" alt="Buy Me A Coffee" style="height: auto !important;width: auto !important;" ></a>

### GPGit
[GPGit][12] is meant to bring GPG to the masses. It is not only a shell script that automates the process of [creating new signed Git releases with GPG][13], but also includes a [quick-start-guide][14] for learning how to use GPG. GPGit integrates perfectly with the [Github Release API][15] for uploading. It can even automatically add a [Keep A Changelog](https://keepachangelog.com/) formatted changelog to the release.

The security status of GNU/Linux projects will be tracked in the [Linux Security Database][16]. If you have any further questions, do not hesitate to [contact me][17] personally. Thanks for your help in making GNU/Linux projects more secure by using GPG signatures.

[1]: https://github.com/NicoHood/gpgit#12-key-generation
[2]: https://github.com/NicoHood/gpgit#11-strong-unique-secret-passphrase
[3]: https://github.com/NicoHood/gpgit#21-send-gpg-key-to-a-key-server
[4]: https://github.com/NicoHood/gpgit#22-publish-full-fingerprint
[5]: https://github.com/NicoHood/gpgit#31-configure-git-gpg-key
[6]: https://github.com/NicoHood/gpgit#32-commit-signing
[7]: https://github.com/NicoHood/gpgit#33-create-signed-git-tag
[8]: https://github.com/NicoHood/gpgit#42-sign-the-archive
[9]: https://github.com/NicoHood/gpgit#41-create-compressed-archive
[10]: https://github.com/NicoHood/gpgit#43-create-the-message-digest
[11]: https://github.com/NicoHood/gpgit#51-configure-https-download-server
[12]: https://github.com/NicoHood/gpgit
[13]: https://github.com/NicoHood/gpgit#script-usage
[14]: https://github.com/NicoHood/gpgit#gpg-quick-start-guide
[15]: https://github.com/NicoHood/gpgit#52-upload-to-github
[16]: https://github.com/NicoHood/LSD
[17]: http://contact.nicohood.de

# Index
* [Introduction](#introduction)
* [GPGit Documentation](#gpgit-documentation)
* [GPG Quick Start Guide](#gpg-quick-start-guide)

# GPGit Documentation

## Installation

### Arch Linux

GPGit is available as [official Arch Linux distribution package](https://archlinux.org/packages/community/any/gpgit/):

```bash
sudo pacman -S gpgit
# Optional dependencies for Github API uploading
sudo pacman -S curl jq
```

### Debian

First install the following dependencies, then follow the manual installation instruction.

```bash
# Install dependencies
sudo apt-get install bash gnupg2 git tar xz-utils coreutils gawk grep sed util-linux
# Optional dependencies
sudo apt-get install gzip bzip lzip zstd file jq curl
```

### MacOS

First install the following dependencies with [Homebrew](https://brew.sh/), then follow the manual installation instructions.

```bash
# Install dependencies
brew install bash git xz gnu-getopt coreutils
# Install a GPG suite, such as https://gpgtools.org/
brew install --cask gpg-suite
# Optional dependencies
brew install gzip bzip2 lzip zstd jq curl
```

### Manual Installation

#### Dependencies

* bash
* gnupg2
* git
* tar
* xz
* grep
* sed
* gnu awk
* gnu getopt (util-linux)
* gnu date (coreutils)

#### Optional Dependencies

* gzip (Compression option)
* zstd (Compression option)
* bzip (Compression option)
* lzip (Compression option)
* file (Github API upload)
* jq (Github API upload)
* curl (Github API upload)
* shellcheck (Development: `make test`)

#### Installation Instructions

```bash
# Download and verify source
VERSION=1.5.0
wget "https://github.com/NicoHood/gpgit/releases/download/${VERSION}/gpgit-${VERSION}.tar.xz"
wget "https://github.com/NicoHood/gpgit/releases/download/${VERSION}/gpgit-${VERSION}.tar.xz.asc"
gpg2 --keyserver hkps://keyserver.ubuntu.com --recv-keys 97312D5EB9D7AE7D0BD4307351DAE9B7C1AE9161
gpg2 --verify "gpgit-${VERSION}.tar.xz.asc" "gpgit-${VERSION}.tar.xz"

# Extract, install and run GPGit
tar -xf "gpgit-${VERSION}.tar.xz"
sudo make -C "gpgit-${VERSION}" PREFIX=/usr/local install
gpgit --help
```

## Script Usage
The script guides you through all 5 steps of the [GPG quick start guide](#gpg-quick-start-guide). Run GPGit with the tag name as parameter. All other options will get auto detected. When running the script for the first time GPGit runs in interactive (`-i`) mode and guide you through all steps of secure source code signing.

If you add and commit a `CHANGELOG.md` file to your Git with the [Keep a Changelog](https://keepachangelog.com/) format, GPGit will autodetect that file and add the corresponding changelog section to the tag message and Github release notes.

### Sample Usage
![screenshot](img/screenshot.png)

### Parameters and Configuration
```
$ gpgit --help
Usage: gpgit [options] <tagname> [<commit> | <object>]

GPGit 1.5.0 https://github.com/NicoHood/gpgit
A shell script that automates the process of signing Git sources via GPG.

Mandatory arguments:
  <tagname>                The name of the tag to create.

Optional arguments:
  -h, --help               Show this help message and exit.
  -m, --message <msg>      Use the given <msg> as the commit message.
                           If multiple -m options are given, their values are
                           concatenated as separate paragraphs.
  -C, --directory <path>   Run as if GPGit was started in <path> instead of the
                           current working directory.
  -u, --local-user <keyid> Use the given GPG key (same as --signingkey).
  -o, --output <path>      Safe all release assets to the specified <path>.
  -a, --asset              Add additional Github assets, e.g. software bundles.
  -t, --title              Custom Github release title (instead of tag name).
  -p, --pre-release        Flag as Github pre-release.
  -f, --force              Force the recreation of Git tag and release assets.
  -i, --interactive        Run in interactive mode, step-by-step.
      --<option>           Temporary set a 'gpgit.<option>' from config below.
  <commit>, <object>       The object that the new tag will refer to.

Examples:
  gpgit 1.0.0
  gpgit -p -m "First alpha release." 0.1.0 --hash "sha256 sha512"
  gpgit -C git/myproject/ -o /tmp/gpgit -n -m "Internal test release." 0.0.1

Configuration options:
  gpgit.signingkey <keyid>, user.signingkey <keyid>
  gpgit.output <path>
  gpgit.token <token>
  gpgit.compression <xz | gzip | bzip2 | lzip | zstd | zip>
  gpgit.hash <sha512 | sha384 | sha256 | sha1 | md5>
  gpgit.changelog <auto | true | false>
  gpgit.github <auto | true | false>
  gpgit.githubrepo <username/projectname>
  gpgit.project <projectname>
  gpgit.keyserver <keyserver>

Examples:
  git config --global gpgit.output ~/gpgit
  git config --local user.signingkey 97312D5EB9D7AE7D0BD4307351DAE9B7C1AE9161
  git config --local gpgit.compression "xz zip"
```

# GPG Quick Start Guide
GPGit guides you through 5 simple steps to get your software project ready with GPG signatures. Further details can be found below.

1. [Generate a new GPG key](#1-generate-a-new-gpg-key)
    1. [Strong, unique, secret passphrase](#11-strong-unique-secret-passphrase)
    2. [Key generation](#12-key-generation)
2. [Publish your key](#2-publish-your-key)
    1. [Send GPG key to a key server](#21-send-gpg-key-to-a-key-server)
    2. [Publish full fingerprint](#22-publish-full-fingerprint)
    3. [Associate GPG key with Github](#23-associate-gpg-key-with-github)
3. [Use Git with GPG](#3-use-git-with-gpg)
    1. [Configure Git GPG key](#31-configure-git-gpg-key)
    2. [Enble commit signing](#32-enable-commit-signing)
    3. [Create signed Git tag](#33-create-signed-git-tag)
4. [Create a signed release archive](#4-create-a-signed-release-archive)
    1. [Create compressed archive](#41-create-compressed-archive)
    2. [Sign the archive](#42-sign-the-archive)
    3. [Create the message digest](#43-create-the-message-digest)
5. [Upload the release](#5-upload-the-release)
    1. [Configure HTTPS download server](#51-configure-https-download-server)
    2. [Upload to Github](#52-upload-to-github)

## 1. Generate a new GPG key
### 1.1 Strong, unique, secret passphrase
Make sure that your new passphrase for the GPG key meets high security standards. If the passphrase/key is compromised all of your signatures are compromised too.

Here are a few examples how to keep a passphrase strong but easy to remember:
* [Creating a strong password](https://support.google.com/accounts/answer/32040?hl=en)
* [How to Create a Secure Password](https://open.buffer.com/creating-a-secure-password/)
* [Mooltipass](https://www.themooltipass.com/)
* [Keepass](https://keepass.info/), [KeepassXC](https://keepassxc.org/)
* [PasswordCard](https://www.passwordcard.org/en)

### 1.2 Key generation
If you don't have a GPG key yet, create a new one first. You can use RSA (4096 bits) or ECC (Curve 25519) for a strong key. GPG offers you the option to use the most future-proof key algorithm available. Use the most recent version gnupg2, not gnupg1!

Ed25519 ECC GPG keys are still not supported by every software/platform. To generate an RSA key use `rsa4096`  instead of `future-default` as parameter.

**Make sure that your secret key is stored somewhere safe and use a unique strong password.**

##### Example key generation:
```bash
$ gpg2 --quick-generate-key "John Doe <john@doe.com>" future-default default 1y
We need to generate a lot of random bytes. It is a good idea to perform
some other action (type on the keyboard, move the mouse, utilize the
disks) during the prime generation; this gives the random number
generator a better chance to gain enough entropy.
gpg: key 2F8E73B1D445CCD3 marked as ultimately trusted
gpg: revocation certificate stored as '/home/john/.gnupg/openpgp-revocs.d/6718A9A63030E182A86FEE152F8E73B1D445CCD3.rev'
public and secret key created and signed.

pub   ed25519 2017-09-24 [SC] [expires: 2018-09-24]
      6718A9A63030E182A86FEE152F8E73B1D445CCD3
uid                      John Doe <john@doe.com>
sub   cv25519 2017-09-24 [E]
```

The generated key has the fingerprint `6718A9A63030E182A86FEE152F8E73B1D445CCD3` in this example. Share it with others so they can verify your source. [[Read more]](https://wiki.archlinux.org/index.php/GnuPG#Create_a_key_pair)

If you ever move your installation make sure to backup `~/.gnupg/` as it contains the **private key** and the **revocation certificate**. Handle it with care. [[Read more]](https://wiki.archlinux.org/index.php/GnuPG#Revoke_a_key)

## 2. Publish your key

### 2.1 Send GPG key to a key server
To make the public key widely available, upload it to a key server. Now the user can get your key by requesting the fingerprint from the keyserver: [[Read more]](https://wiki.archlinux.org/index.php/GnuPG#Use_a_keyserver)

```bash
# Publish key
gpg2 --keyserver hkps://keyserver.ubuntu.com --send-keys <fingerprint>

# Import key
# Alternative keyserver: hkps://hkps.pool.sks-keyservers.net
gpg2 --keyserver hkps://keyserver.ubuntu.com --recv-keys <fingerprint>
```

### 2.2 Publish full fingerprint
To make it easy for everyone else to find your key it is crucial that you publish the [**full fingerprint**](https://lkml.org/lkml/2016/8/15/445) on a trusted platform, such as your website or Github. To give the key more trust other users can sign your key too. [[Read more]](https://wiki.debian.org/Keysigning)

### 2.3 Associate GPG key with Github
To make Github display your commits as "verified" you also need to add your public [GPG key to your Github profile](https://github.com/settings/keys). [[Read more]](https://docs.github.com/github/authenticating-to-github/adding-a-new-gpg-key-to-your-github-account)

```bash
# List keys + full fingerprint
gpg2 --list-secret-keys --keyid-format LONG

# Generate public key
gpg2 --armor --export <fingerprint>

# If you have multiple uids or signatures you can minimize the output:
gpg2 --armor --export --export-filter keep-uid="uid =~ <email>" --export-options export-minimal <fingerprint>
```

## 3. Use Git with GPG
### 3.1 Configure Git GPG key
In order to make Git use your GPG key you need to set the default signing key for Git. [[Read more]](https://help.github.com/articles/telling-git-about-your-gpg-key/)

```bash
# List keys + full fingerprint
gpg2 --list-secret-keys --keyid-format LONG

git config --global user.signingkey <fingerprint>
```

### 3.2 Enable commit signing
To verify the Git history, Git commits needs to be signed. You can manually sign commits or enable it by default for every commit. It is recommended to globally enable Git commit signing. [[Read more]](https://help.github.com/articles/signing-commits-using-gpg/)

```bash
git config --global commit.gpgsign true
```

### 3.3 Create signed Git tag

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
This changlog uses the [ISO 8601 date format](https://www.iso.org/iso-8601-date-and-time-format.html) of (YYYY-MM-DD).

## [Unreleased]

### Changed

* Added project name (and tag) to GitHub release title

### Fixed

* Improved remote name detection and throw an error if that is not possible.
* Fixed error with unbound color variables and improved error reporting of it.

## [1.5.0] - 2021-07-11

### Added

* zstd compression support
* Added .editorconfig
* Added key expire readme tutorial [#26](https://github.com/NicoHood/gpgit/issues/26)
* Added `--version` option
* Added `--asset` option [#32](https://github.com/NicoHood/gpgit/issues/32)
* Added `--title` option [#28](https://github.com/NicoHood/gpgit/issues/28)

### Fixed

* Fixed Readme Links

## [1.4.1] - 2021-01-31

### Changed

* Update license date and add license to script

### Fixed

* Fixed Github API json escaping issue
* Added depcrecated warning about `--no-github` parameter

## [1.4.0] - 2021-01-29

### Added

* Added CHANGELOG.md file with [Keep A Changelog format](https://keepachangelog.com)
* Added support for automatic [Keep A Changelog format](https://keepachangelog.com) parsing
* Added [additional quoting for command substitution](https://unix.stackexchange.com/q/118433)
* Detect default github branch automatically when tagging a specific commit
* Added [ShellCheck](https://www.shellcheck.net/) makefile `test` target
* Added detection of used remote/upstream.
* Show github release link after uploading.

### Changed

* Explicitly use annotated git tag using `-a`
* Improved grep handing using -F option
* Add a better error message if signing git tag failed.
* Disable interactive mode for first run only when script finishes properly #20
* Improved compression and hash bash array handling
* Changed default keyserver to `hkps://keyserver.ubuntu.com` as `hkps://pgp.mit.edu` is offline most of the time.

### Deprecated
* Changed `-n`, `--no-github` parameters to `--github false`.

### Removed
* Remove not required exit command
* Remove unused MAGENTA and CYAN colors

### Fixed

* Fixed reading private repository information by always specifying the token
* Fix `--no-github` parameter
* Fixed `--force` option for github releases #24
* Added support for BSD based systems #19 (thanks @WoLpH)

## [1.3.4] - 2020-03-23

### Added

* Add additional keyserver error message #18

### Fixed
* Fix crash when using a none github repo

## [1.3.3] - 2018-10-31

### Fixed
* Create parent directories if not existant.

## [1.3.2] - 2018-04-15

### Fixed
* Change directory to git root path, so "git archive" is working properly

## [1.3.1] - 2018-03-14

### Added
* Added -f, --force option
* Added version identifier in default tag message
* Added option to specify the commit/object to tag

### Changed

* Improved hash algorithm usage for non linux systems
* Improved keyid parameter to fit better with git tag and gpg

### Fixed

* Fixed shebang for non linux systems
* Fixed #17 git config reading for commit.gpgsign setting

## [1.3.0] - 2018-01-24

### Added

* Added environment variable and git config support
* Added color output options
* Added zip support
* Added support for multiple compression/hash algorithms

### Changed

* Reworked bash script completely
* Simplified parameters
* Reduced output verbosity
* Generate archive from local git source rather than downloading it from github
* Do less unnecessary error checking, but simplify the code instead
* Create signatures with strongest hash algorithm
* Use ECC keys for GPG key generation if available

### Fixed
* Fix pushing tag if a branch with the same name also exists

## [2.0.7] - 2017-06-27

**Note: The 2.x.x series is written in Python, but was discontinued because of multiple issues with the python libraries. A Bash script is much easier to integrate and set up, so development was refocused on that.**

### Added

* Switch to Python3 from bash
* New user interface with preview
* More verification
* Better GPG usage
* More parameters
* Configurable settings via git config
* Better error traces
* Resigning a tag is now possible
* General improvements
* New logo
* Improved documentation

## [1.2.0] - 2017-04-24

### Added
* Trap on errors
* Detect gpg2

### Fixed
* Fix Git tags pull/push
* Small code fixes #3 (Thanks @cmaglie)

## [1.1.2] - 2017-01-22

### Fixed
* Fixed Github uploading name

## [1.1.1] - 2017-01-17

### Added
* Verify existing signatures
* Added upload to Github functionality

### Changed
* Only allow secure GPG keys

## [1.1.0] - 2017-01-13

### Added
* Added online source download
* Added source verification
* Added multiple compression algorithms
* Added multiple sha algorithms

### Changed
* Updated Readme

### Fixed

* Minor fixes

## [1.0.0] - 2017-01-07

### Added
- First release with all functions working except the uploading

### Changed
- Merged all scripts into gpgit.sh

## [0.1.0] - 2016-12-16

### Added
- Initial release of the software

[Unreleased]: https://github.com/NicoHood/gpgit/compare/1.5.0...HEAD
[1.5.0]: https://github.com/NicoHood/gpgit/compare/1.4.1...1.5.0
[1.4.1]: https://github.com/NicoHood/gpgit/compare/1.4.0...1.4.1
[1.4.0]: https://github.com/NicoHood/gpgit/compare/1.3.4...1.4.0
[1.3.4]: https://github.com/NicoHood/gpgit/compare/1.3.3...1.3.4
[1.3.3]: https://github.com/NicoHood/gpgit/compare/1.3.2...1.3.3
[1.3.2]: https://github.com/NicoHood/gpgit/compare/1.3.1...1.3.2
[1.3.1]: https://github.com/NicoHood/gpgit/compare/1.3.0...1.3.1
[1.3.0]: https://github.com/NicoHood/gpgit/compare/1.2.0...1.3.0
[2.0.7]: https://github.com/NicoHood/gpgit/releases/tag/2.0.7
[1.2.0]: https://github.com/NicoHood/gpgit/compare/1.1.2...1.2.0
[1.1.2]: https://github.com/NicoHood/gpgit/compare/1.1.1...1.1.2

gpgithub.sh
#!/bin/bash

# Stop on errors
set -e -u -o pipefail

# Avoid any encoding problems
export LANG=C

PROGNAME=$(basename "$0")
ARGS=( "$@" )

usage()
{
    echo "Usage: ${PROGNAME} <tag> [options]"
    echo
    echo 'Mandatory parameters:'
    echo '<tag>           Tagname'
    echo
    echo 'Actions:'
    echo '-h --help       Show this help message'
    echo
    echo 'Options:'
    echo '-o, --output    The output path of the .tar.gz, .sig and sha512'
    echo '                Default: "git rev-parse --show-toplevel)/archive"'
    echo '-u, --username  Username of the user. Used for GPG key generation.'
    echo '                Default: git config user.name'
    echo '-e, --email     Email of the user. Used for GPG key generation.'
    echo '                Default: "git config user.email"'
    echo '-p, --project   The name of the project. Used for archive geneation.'
    echo "                Default: \"git config --local remote.origin.url \\"
    echo "                           | sed -n \'s#.*/\([^.]*\)\.git#\1#p\'\""
	echo '-g, --gpg       Specify (full) GPG fingerprint to use for signing.'
    echo '                Default: "git config user.signingkey"'
	echo '-m, --message   Specify the tag message.'
	echo '                Default: "Release <tag>"'
	echo '-y, --yes       Assume "yes" on all questions.'
}

################################################################################
# Functions
################################################################################

# Check if messages are to be printed using color
unset ALL_OFF BOLD BLUE GREEN RED YELLOW
if [[ -t 2 ]]; then
    # prefer terminal safe colored and bold text when tput is supported
    if tput setaf 0 &>/dev/null; then
        ALL_OFF="$(tput sgr0)"
        BOLD="$(tput bold)"
        BLUE="${BOLD}$(tput setaf 4)"
        GREEN="${BOLD}$(tput setaf 2)"
        RED="${BOLD}$(tput setaf 1)"
        YELLOW="${BOLD}$(tput setaf 3)"
    else
        ALL_OFF="\e[1;0m"
        BOLD="\e[1;1m"
        BLUE="${BOLD}\e[1;34m"
        GREEN="${BOLD}\e[1;32m"
        RED="${BOLD}\e[1;31m"
        YELLOW="${BOLD}\e[1;33m"
    fi
fi
readonly ALL_OFF BOLD BLUE GREEN RED YELLOW

msg() {
    local mesg=$1; shift
    printf "${GREEN}==>${ALL_OFF}${BOLD} ${mesg}${ALL_OFF}\n" "$@" >&2
}

msg2() {
    local mesg=$1; shift
    printf "${BLUE}  ->${ALL_OFF}${BOLD} ${mesg}${ALL_OFF}\n" "$@" >&2
}

plain() {
    local mesg=$1; shift
    printf "${BOLD}    ${mesg}${ALL_OFF}\n" "$@" >&2
}

warning() {
    local mesg=$1; shift
    printf "${YELLOW}==> WARNING:${ALL_OFF}${BOLD} ${mesg}${ALL_OFF}\n" "$@" >&2
}

error() {
    local mesg=$1; shift
    printf "${RED}==> ERROR:${ALL_OFF}${BOLD} ${mesg}${ALL_OFF}\n" "$@" >&2
}

info() {
    local mesg=$1; shift
    printf "${YELLOW}[!]:${ALL_OFF}${BOLD} ${mesg}${ALL_OFF}\n" "$@" >&2
}

gpgithub_yesno() {
	[[ "${config[YES]}" == true ]] && return
    read -rp "${BOLD}    Continue? [Y/n]${ALL_OFF}" yesno
    if [[ "${yesno}" != [Yy]"es" && "${yesno}" != [Yy] && -n "${yesno}" ]]; then
        warning "Aborted by user"
        exit 0
    fi
}

################################################################################
# Parameters
################################################################################

# Check if inside a git folder
if [[ "$(git rev-parse --is-inside-work-tree)" != "true" ]]; then
    error "Not a git repository."
fi

# Check input param number
if [[ $# -lt 1 ]]; then
    error "Usage: ${PROGNAME} <tag>" 1>&2
    plain "Use --help for more information."
    exit 1
fi

# Print help
if [[ "$1" == "-h" || "$1" == "--help" ]]; then
    usage 1>&2
    exit 0
fi

# Set default values in config array
typeset -A config
config=(
    [TAG]="$1"
    [OUTPUT]="$(git rev-parse --show-toplevel)/archive"
    [USERNAME]="$(git config user.name)"
    [EMAIL]="$(git config user.email)"
    [PROJECT]="$(git config --local remote.origin.url | sed -n 's#.*/\([^.]*\)\.git#\1#p')"
    [GPG]="$(git config user.signingkey)"
	[MESSAGE]="Release $1"
	[YES]=false
)
shift

# Parse input params an ovrwrite possible default or config loaded options
GETOPT_ARGS=$(getopt -o "ho:u:e:p:g:m:y" \
            -l "help,output:,username:,email:,project:,gpg:,message:,yes"\
            -n "$PROGNAME" -- "$@")
eval set -- "$GETOPT_ARGS"

# Handle all params
while true ; do
    case "$1" in
        # Options
        -o|--output)
            config[OUTPUT]="$2"
            shift
            ;;
        -u|--username)
            config[USERNAME]="$2"
            shift
            ;;
        -e|--email)
            config[EMAIL]="$2"
            shift
            ;;
        -p|--project)
            config[PROJECT]="$2"
            shift
            ;;
		-g|--gpg)
            config[GPG]="$2"
            shift
            ;;
		-m|--message)
			config[MESSAGE]="$2"
			shift
			;;
		-y|--yes)
			config[YES]=true
			;;
        # Internal
        -h|--help)
            usage 1>&2
            exit 0
            ;;
        --)
            # No more options left.
            shift
            break
           ;;
        *)
            error "Internal error!"
            exit 1
            ;;
    esac
    shift
done

################################################################################
msg "1. Generate new GPG key"
################################################################################

# Check for existing key
if [[ -z "${config[GPG]}" ]]; then
    if gpg --list-secret-keys | grep uid | grep -v -q revoked; then
        error "GPG seems to be already configured on your system but git is not."
        plain "Please use gpg --list-secret-keys to show existing keys."
        plain "Afterwards set the key with git config --global user.signingkey <key>."
        plain "See the readme for more information."
        exit 1
    else
        plain "Generating an RSA 4096 GPG key for ${config[USERNAME]} <${config[EMAIL]}> valid for 3 years."
        gpgithub_yesno

        # Generate ECC key command (currently not supported by Github)
        #gpg --quick-generate-key "testuser (comment) <name@mail.com>" future-default default 3y

        # Generate RSA key command
        # https://www.gnupg.org/documentation/manuals/gnupg/Unattended-GPG-key-generation.html
        # gpg: revocation certificate stored as '/tmp/tmp.81v03YSxmI/openpgp-revocs.d/F4EDF85EFF03D746D17094D3C28B8F6BCCDF8671.rev'
        config[GPG]="$(gpg --batch --generate-key <( cat << EOF
            Key-Type: RSA
            Key-Length: 4096
            Key-Usage: cert sign auth
            Subkey-Type: RSA
            Subkey-Length: 4096
            Subkey-Usage: encrypt
            Name-Real: ${config[USERNAME]}
            #Name-Comment: Generated with gpgit
            Name-Email: ${config[EMAIL]}
            Expire-Date: 3y
            # Preferences: TODO https://security.stackexchange.com/questions/82216/how-to-change-default-cipher-in-gnupg-on-both-linux-and-windows
            %ask-passphrase
            %echo We need to generate a lot of random bytes. It is a good idea to perform
            %echo some other action (type on the keyboard, move the mouse, utilize the
            %echo disks) during the prime generation; this gives the random number
            %echo generator a better chance to gain enough entropy.
            %commit
            %echo Key generation finished
EOF
        ) 2>&1 | tee -a /dev/fd/2 | grep "revocation certificate stored as " \
               | sed 's,.*/\(.*\).rev.*,\1,')"
        NEW_GPG_KEY=true

        # Print new fingerprint
        plain "Your new GPG fingerprint is: ${config[GPG]}"
        gpg -u "${config[GPG]}" --list-secret-keys --keyid-format LONG
    fi
else
    plain "Key already generated. Using key: ${config[GPG]}"
	NEW_GPG_KEY=false

	# Check if the full fingerprint is used
	if [[ ${#config[GPG]} -ne 40 ]]; then
		error "Please specify the full fingerprint."
		exit 1
	fi

    # Check if key exists
    if ! gpg --keyid-format LONG --list-secret-keys "0x${config[GPG]}"; then
        error "This GPG key is not known on this system."
        plain "Check your git config or your GNUPGHOME variable."
        exit 1
    fi
fi

################################################################################
msg "2. Publish your key"
################################################################################

# Check if key was just created
if [[ "${NEW_GPG_KEY}" = true ]]; then
    # Refresh setting
    config[GPG]="$(git config --global user.signingkey)"

    # Upload key
    msg2 "2.1 Submit your key to a key server"
    plain "Uploading key ${config[GPG]} to hkps://hkps.pool.sks-keyservers.net"
    gpgithub_yesno
    gpg --keyserver hkps://hkps.pool.sks-keyservers.net --send-keys "${config[GPG]}"

    # Generate public key
    msg2 "2.2 Associate GPG key with github"
    plain "Please visit Github and add the following GPG key to your profile."
    plain "https://github.com/settings/keys"
    gpgithub_yesno
    gpg --armor --export "${config[GPG]}"

    msg2 "2.3 Publish your fingerprint"
    plain "Publish your GPG fingerprint (${config[GPG]}) on your project site."
    plain "Also see https://wiki.debian.org/Keysigning"
    gpgithub_yesno
else
	plain "Assuming key was already publish with its generation. If not please do so."
fi

################################################################################
msg "3. Usage of GPG by git"
################################################################################

# Differenciate between new created key and (temporary) different key
if [[ "${NEW_GPG_KEY}" = true ]]; then
    GIT_CONFIG="global"
else
    GIT_CONFIG="local"
fi

#  3.1 Configure git GPG key
msg2 "3.1 Configure git GPG key"
if [[ "${config[GPG]}" != "$(git config user.signingkey)" ]]; then
    # If the key differs from the local>global>system configured key, set it locally
	plain "Git is not configured with this key."
    plain "Configuring ${GIT_CONFIG} git settings with your GPG key."
    gpgithub_yesno
    git config --"${GIT_CONFIG}" user.signingkey "${config[GPG]}"
else
    plain "Git already configured with your GPG key"
fi

# Check if commit signing is enabled for this repo and ask for a switch if not
msg2 "3.2 Commit signing"
if [[ $(git config commit.gpgsign) != true ]]; then
    warning "Commit signing is disabled. Will enable it now ${GIT_CONFIG}ly."
    gpgithub_yesno
    git config --"${GIT_CONFIG}" commit.gpgsign true
else
    plain "Commit signing already enabled."
fi

# Refresh tags
msg2 "3.3 Create signed git tag"
plain "Refreshing tags from upstream."
gpgithub_yesno
git pull --tags

# Check if tag exists
if ! git tag | grep "^${config[TAG]}$" -q; then
    # Check if every added file has been commited
    if ! git diff --cached --exit-code; then
        warning "You have added new changes but did not commit them yet."
        gpgithub_yesno
    fi

    # Create new tag if not existant
    plain "Creating signed tag ${config[TAG]} and pushing it to the remote git."
    gpgithub_yesno

    # Create and push new git tag
    git tag -s "${config[TAG]}" -m "${config[MESSAGE]}"
    git push --tags
else
    plain "Tag ${config[TAG]} already exists."
fi

################################################################################
msg "4. Creation of a signed compressed release archive"
################################################################################

# Check if output path exists and ask for creation
if [[ ! -d "${config[OUTPUT]}" ]]; then
    plain "Output path does not exist. Create ${config[OUTPUT]} ?"
    gpgithub_yesno
    mkdir -p "${config[OUTPUT]}"
fi

# Build .tar path
config[TAR]="${config[OUTPUT]}/${config[PROJECT]}-${config[TAG]}.tar"

# Create .tar.xz archive with maximum compression if not existant
msg2 "4.1 Create compressed archive"
if [[ -f "${config[TAR]}.xz" ]]; then
    plain "Archive ${config[TAR]}.xz already exists."
    gpgithub_yesno
else
    plain "Creating release archive file ${config[TAR]}.xz"
	gpgithub_yesno
    git archive --format=tar --prefix "${config[PROJECT]}-${config[TAG]}/" "${config[TAG]}" | xz -9 > "${config[TAR]}.xz"
fi

# Create sha512 of the .tar.xz
msg2 "4.2 Create the message digest"
gpgithub_yesno
sha512sum "${config[TAR]}.xz" > "${config[TAR]}.xz.sha512"

# Sign .tar.xz if not existant
msg2 "4.3 Sign the sources"
if [[ -f "${config[TAR]}.xz.sig" ]]; then
    plain "Signature for ${config[TAR]}.xz already exists."
    gpgithub_yesno
else
    plain "Creating signature for file ${config[TAR]}.xz"
	gpgithub_yesno
    gpg --local-user "${config[GPG]}" --output "${config[TAR]}.xz.sig" --armor --detach-sign "${config[TAR]}.xz"
fi

################################################################################
msg "5. Upload the release"
################################################################################

#TODO
msg2 "5.1 Github"
plain "TODO"
# Create new Github release if not existant
# Upload files to Github
#
# # Create github release and upload the signature
# # http://www.barrykooij.com/create-github-releases-via-command-line/
# # https://developer.github.com/v3/repos/releases/
# # https://developer.github.com/changes/2013-09-25-releases-api/
# read -rsp "Enter your Github token (Github->Settings->Personal access tokens; public repo access):" TOKEN
# API_JSON=$(printf '{"tag_name": "%s","target_commitish": "%s","name": "%s","body": "Release %s","draft": false,"prerelease": false}' "${TAG}" "${BRANCH}" "${TAG}" "${TAG}")
# if ! RESULT=$(curl --data "$API_JSON" "https://api.github.com/repos/${USER_NAME}/${PROJECT_NAME}/releases" \
# -H "Accept: application/vnd.github.v3+json" -H "Authorization: token ${TOKEN}" ); then
#     echo "Error: Uploading failed. Release already exists or token is wrong?" 1>&2
#     exit 1
# fi
# RELEASE_ID=$(echo "${RESULT}" | grep '^  "id": ' | tr -dc '[:digit:]')
#
# if ! curl "https://uploads.github.com/repos/${USER_NAME}/${PROJECT_NAME}/releases/${RELEASE_ID}/assets?name=${PROJECT_NAME}-${TAG}.tar.gz.sig" \
# -H "Content-Type: application/pgp-signature" \
# -H "Accept: application/vnd.github.v3+json" \
# -H "Authorization: token ${TOKEN}" \
# --data-binary @"${TAR_GZ}.sig"; then
#     echo "Error: Uploading failed. Release already exists or token is wrong?" 1>&2
#     exit 1
# fi
#
# if ! curl "https://uploads.github.com/repos/${USER_NAME}/${PROJECT_NAME}/releases/${RELEASE_ID}/assets?name=${PROJECT_NAME}-${TAG}.tar.gz.sha512" \
# -H "Content-Type: text/sha512" \
# -H "Accept: application/vnd.github.v3+json" \
# -H "Authorization: token ${TOKEN}" \
# --data-binary @"${TAR_GZ}.sha512"; then
#     echo "Error: Uploading failed. Release already exists or token is wrong?" 1>&2
#     exit 1
# fi
